include/envoy/network/listen_socket.h

```diff
@@ -34,5 +34,65 @@ class ListenSocket {
 typedef std::unique_ptr<ListenSocket> ListenSocketPtr;
 typedef std::shared_ptr<ListenSocket> ListenSocketSharedPtr;
 
+/**
+ * A socket passed to a connection. For server connections this represents the accepted socket, and
+ * for client connections this represents the socket being connected to a remote address.
+ *
+ * TODO(jrajahalme): Hide internals (e.g., fd) from listener filters by providing callbacks filters
+ * may need (set/getsockopt(), peek(), recv(), etc.)
+ */
+class ConnectionSocket {
+public:
+  virtual ~ConnectionSocket() {}
+
+  /**
+   * @return the local address of the socket.
+   */
+  virtual const Address::InstanceConstSharedPtr& localAddress() const PURE;
+
+  /**
+   * @return the remote address of the socket.
+   */
+  virtual const Address::InstanceConstSharedPtr& remoteAddress() const PURE;
+
+  /**
+   * Set the local address of the socket. On accepted sockets the local address defaults to the
+   * one at which the connection was received at, which is the same as the listener's address, if
+   * the listener is bound to a specific address.
+   *
+   * @param local_address the new local address.
+   * @param restored a flag marking the local address as being restored to a value that is
+   *        different from the one the socket was initially accepted at. This should only be set
+   *        to 'true' when restoring the original destination address of a connection redirected
+   *        by iptables REDIRECT. The caller is responsible for making sure the new address is
+   *        actually different when passing restored as 'true'.
```

Comment:
Can you also give an example of when someone would call this function but set `restored` to false? I see multiple examples in this PR but just from reading the interfaces it's not completely clear when one would do that.

Reply:
I reverted the meaning of this boolean and renamed it as "hand_off_restored_destinations", should be clearer now, see the new commit.